src/Type/UnionType.php

```diff
@@ -149,6 +154,42 @@ public function isSuperTypeOf(Type $otherType): TrinaryLogic
 		return $result;
 	}
 
+	private function isSuperTypeOfIntegerRange(IntegerRangeType $otherType): TrinaryLogic
+	{
+		$min = $otherType->getMin();
+		$max = $otherType->getMax();
+		if ($min === null || $max === null) {
+			throw new ShouldNotHappenException();
+		}
+
+		$matchingConstantIntegers = 0;
+		$results = [];
+
+		foreach ($this->types as $innerType) {
+			if ($innerType instanceof ConstantIntegerType) {
+				if ($innerType->getValue() >= $min && $innerType->getValue() <= $max) {
+					$results[] = TrinaryLogic::createMaybe();
+					$matchingConstantIntegers++;
+					continue;
+				}
+				return TrinaryLogic::createNo();
+			}
+
+			$result = $innerType->isSuperTypeOf($otherType);
+			if ($result->yes()) {
+				return $result;
+			}
```

Comment:
I think the logic here is contrary. 

in line 175 you return 'No' as soon as you find one constant integer which is not part of the range (no matter what other elements are in the union), but in line 180 you return 'Yes' as soon as you find one element within the union which is a super type of the integer range - no matter which other elements are in the union?

Reply:
Good catch - I've simplified the implementation a bit to only do the actual range check here